### ISSUE TITLE
Needless META-INF/resources and web-fragment.xml mounts

### ISSUE THREAD

**Issue by gregw** [eclipse/jetty.project#10164]
**Jetty version(s)**
12

**Enhancement Description**

If you deploy the ee10-demo-async-rest
and dump the server, you see that the ResourceFactory for the context contains a `META-INF/resources` and `META-INF/web-fragment.xml` entries for every jar file, even though they do not exist.

Either we should not mount these resources if they do not exist.... or if we are caching non-existence, we should exclude them from a dump.

```
|     += oejur.ResourceFactoryInternals$LifeCycle@22d7b4f8{STARTED} - STARTED
|     |  +> mounts size=30
|     |     +> Mount[uri=jar:file:///home/gregw/src/jetty-12.0.x/jetty-home/target/jetty-home/lib/jakarta.servlet-api-6.0.0.jar!/,root=jar:file:///home/gregw/src/jetty-12.0.x/jetty-home/target/jetty-home/lib/jakarta.servlet-api-6.0.0.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///home/gregw/src/jetty-12.0.x/jetty-home/target/jetty-home/lib/jakarta.servlet-api-6.0.0.jar!/,root=jar:file:///home/gregw/src/jetty-12.0.x/jetty-home/target/jetty-home/lib/jakarta.servlet-api-6.0.0.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///home/gregw/src/jetty-12.0.x/jetty-home/target/jetty-home/lib/jakarta.servlet-api-6.0.0.jar!/,root=jar:file:///home/gregw/src/jetty-12.0.x/jetty-home/target/jetty-home/lib/jakarta.servlet-api-6.0.0.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-alpn-client-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-alpn-client-12.0.0-SNAPSHOT.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-alpn-client-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-alpn-client-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-alpn-client-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-alpn-client-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-client-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-client-12.0.0-SNAPSHOT.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-client-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-client-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-client-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-client-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-ee10-demo-async-rest-jar-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-ee10-demo-async-rest-jar-12.0.0-SNAPSHOT.jar!/META-INF/resources/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-ee10-demo-async-rest-jar-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-ee10-demo-async-rest-jar-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-ee10-demo-async-rest-jar-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-ee10-demo-async-rest-jar-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-http-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-http-12.0.0-SNAPSHOT.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-http-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-http-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-http-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-http-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-io-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-io-12.0.0-SNAPSHOT.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-io-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-io-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-io-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-io-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-slf4j-impl-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-slf4j-impl-12.0.0-SNAPSHOT.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-slf4j-impl-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-slf4j-impl-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-slf4j-impl-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-slf4j-impl-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-12.0.0-SNAPSHOT.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-ajax-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-ajax-12.0.0-SNAPSHOT.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-ajax-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-ajax-12.0.0-SNAPSHOT.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-ajax-12.0.0-SNAPSHOT.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/jetty-util-ajax-12.0.0-SNAPSHOT.jar!/]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/slf4j-api-2.0.7.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/slf4j-api-2.0.7.jar!/META-INF/resources]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/slf4j-api-2.0.7.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/slf4j-api-2.0.7.jar!/META-INF/web-fragment.xml]
|     |     +> Mount[uri=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/slf4j-api-2.0.7.jar!/,root=jar:file:///tmp/jetty-0_0_0_0-8080-ee10-demo-async-rest_war-_ee10-demo-async-rest-any-18389751907668110065/webapp/WEB-INF/lib/slf4j-api-2.0.7.jar!/]
|     +- org.eclipse.jetty.util.resource.ResourceFactory$2@4b8729ff

```

**Comment by gregw** [comment]
Most jar files are not going to have either `META-INF/resources` nor `web-fragments.xml` in them.  It feels highly inefficient that we are mounting all these jars that will already have been opened by the class loader.

Could we use something like `context.getClassLoader().getResources(String)` to find all the fragments and resources without needing to mount the jar files just to check?   I know it gives a horrible `Enumeration` of `URL`s, but we could at least use that as a cheap existence check before mounting the jar ourselves?

At the very very least, can't we unmount the jars that have nothing that we use so we can free the memory used to load them?

@janbartel when we scan a jar for annotations, do we use these mounted resources or does ASM do it's own thing?

**Comment by janbartel** [comment]
When we scan a `Resource` that is a jar we now use it as a `Path` and do a try-with-resources like this:

```java
 try (Stream<Path> classStream = Files.walk(path))
```

**Comment by gregw** [comment]
Those Paths are probably from a ZipFS, so probably mounted.

Hmmm but we don't end up with all those classes mounted on the resource factory? So something different is happening.

Either way, we need to see if we can be more efficient here

**Comment by janbartel** [comment]
NB If we want to get rid of the `Resources` that we keep around (in the context `MetaData`), we need a different mechanism that remembers the ordering and exclusions of jars on the container classpath and the webapp classpath.

**Comment by lorban** [comment]
Actually, each jar file is only mounted once, but there is one mount object created for each `Resource` that is requested via `ResourceFactory.newResource()` to make a poor man's counter; that could be optimized to use a real counter instead so that you would only see 10 mounts in your example.

If we don't need to keep the resources around after the jars have been scanned, we should change the scope of the `ResourceFactory`, for instance by creating a `Closeable` one for the duration of the scan instead of using the one bound to `WebAppContext`.

**Comment by gregw** [comment]
I still think we should investigate if we can get resources from the classloader.  Even 1 mount means that the jar is expanded into memory.

**Comment by joakime** [comment]
> Even 1 mount means that the jar is expanded into memory.

ZipFs works via the `Files.newByteChannel(zipfile, READ)` mechanism (a `SeekableByteChannel`)
It doesn't load the entire jar in memory.
It will, however, maintain the zip/jar central directory in memory (essentially the map of named entries to their locations in the zip/jar)

**Comment by gregw** [comment]
So that's good....

But still I don't understand why there are entries for resources that do not exist?

**Comment by lorban** [comment]
That's an inefficiency of the current implementation: multiple `Mount` objects are created as they lack a resource counter. But they all point to a single real mount.

`ResourceFactory.newResource()` was called for `jar:file:...x.jar!/META-INF/resources`, `jar:file:...x.jar!/META-INF/web-fragment.xml` and `jar:file:...x.jar!/` so the jar file is mounted once, 3 `Mount` objects are created, and once all 3 `Mount` instances are closed the jar gets unmounted. We could use a single `Mount` with a resource counter instead.

**Comment by joakime** [comment]
@lorban the Mount logic in HEAD seems sane.

I added the a testcase to prove this out in PR #10170 (a testcase that passes btw)

Copy/Pasted here ...

``` java
    /**
     * When mounting multiple points within the same JAR, only
     * 1 mount should be created, but have reference counts
     * tracked separately.
     */
    @Test
    public void testMountByJarName()
    {
        Path jarPath = MavenPaths.findTestResourceFile("jar-file-resource.jar");
        URI uriRoot = URI.create("jar:" + jarPath.toUri().toASCIIString() + "!/"); // root
        URI uriRez = URI.create("jar:" + jarPath.toUri().toASCIIString() + "!/rez/"); // dir
        URI uriDeep = URI.create("jar:" + jarPath.toUri().toASCIIString() + "!/rez/deep/"); // dir
        URI uriZzz = URI.create("jar:" + jarPath.toUri().toASCIIString() + "!/rez/deep/zzz"); // file

        try (ResourceFactory.Closeable resourceFactory = ResourceFactory.closeable())
        {
            Resource resRoot = resourceFactory.newResource(uriRoot);
            Resource resRez = resourceFactory.newResource(uriRez);
            Resource resDeep = resourceFactory.newResource(uriDeep);
            Resource resZzz = resourceFactory.newResource(uriZzz);

            assertThat(FileSystemPool.INSTANCE.mounts().size(), is(1));
            int mountCount = FileSystemPool.INSTANCE.getReferenceCount(uriRoot);
            assertThat(mountCount, is(4));
        }
    }
```

**Comment by joakime** [comment]
Looks like the dump is showing the tracked mounts of the ResourceFactory, not the tracked mounts of the FileSystemPool.

The naming in the dump should be cleaned up.
Perhaps "references" instead of "mounts", and only showing the URI used via `ResourceFactory.newResource()` in the output?

The excess entries is a different effort, I'll work on that too.

**Comment by gregw** [comment]
Maybe only dump "references" that exist?

**Comment by lorban** [comment]
I agree that the `Mount` class is misnamed, `Reference` is a bit too generic, but still a better name.

Am I overlooking something, or shouldn't there be just a single `Reference` for _all_ `Resource`s created by one `ResourceFactory`, but one `Reference` by `ResourceFactory`?

**Comment by joakime** [comment]
Also pushed a new PR to fix the usages of `Resource` in the `MetaInfConfiguration` class.
See PR #10171

**Comment by joakime** [comment]
Work in branch https://github.com/jetty/jetty.project/compare/fix/12.0.x/metainf-overhaul will be presented as a PR for 12.0.4

**Comment by joakime** [comment]
Closing as completed already.
